### PR TITLE
Associate guest checkout with account if exists

### LINKED ISSUE #12443: Associate guest checkout with an account if it exists
### What I'm trying to achieve
I have an active customer account at Saleor. When I place an order as a guest but use my account's email, the order is not associated with the account. If the email is confirmed (email confirmation is enabled), my order should be automatically associated with my account.

### Describe a proposed solution
Adjust the `complete_checkout` function to assign the user account if the checkout's email is a valid customer account in Saleor.

### Other solutions I've tried and won't work
…

### PR DESCRIPTION
**Pull request by FremahA**
I want to merge this change because it fixes associating guest checkout emails with an account if it exists.
It fixes #12443

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [x ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/checkout/tests/test_associate_checkout_with_account.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        discounts=None,
+        user=user,
+        app=app
+    )
+
+    # assert that the order is associated with the correct user
+    if checkout_info.user:
+        assert order.user == customer_user
+    else:
+        assert order.user is None
```
Comment (fowczarek) [inline saleor/checkout/tests/test_associate_checkout_with_account.py:53]: Missing new line on the end of the file.

File: saleor/checkout/tests/test_associate_checkout_with_account.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        lines=lines,
+        payment_data={},
+        store_source=False,
+        discounts=None,
+        user=user,
+        app=app
+    )
+
+    # assert that the order is associated with the correct user
+    if checkout_info.user:
```
[the remark this answers is not in the document]
Comment (FremahA) [reply saleor/checkout/tests/test_associate_checkout_with_account.py:50]: @fowczarek Thank you for your review and feedback on my pull request. I will make the necessary changes with these guidelines.
Please let me know if you have any additional suggestions or concerns.

**Review by fowczarek — COMMENTED** [review]
Could you fix the remarks from CI? 
https://github.com/saleor/saleor/actions/runs/4938729730/jobs/8842626974?pr=12758

**Comment by kadewu** [comment]
According to @fowczarek comment you can check contributing guide, especially: [coding style](https://docs.saleor.io/docs/3.x/developer/community/contributing#coding-style)

**Review by fowczarek — APPROVED** [review]

**Review by maarcingebala — COMMENTED** [review]
We should also have a test that checks that passing a valid user instance, but inactive, doesn't associate the checkout.

File: saleor/checkout/tests/test_associate_checkout_with_account.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        ("test@example.com", MarkAsPaidStrategy.TRANSACTION_FLOW),
+        ("test@example.com", MarkAsPaidStrategy.PAYMENT_FLOW),
+    ],
+)
+def test_associate_guest_checkout_with_account_if_exists_with_inactive_user(
+    email, paid_strategy, app, address, checkout, customer_user
+):
+    # set the checkout email
+    checkout.email = email
+    checkout.save()
```
Comment (fowczarek) [inline saleor/checkout/tests/test_associate_checkout_with_account.py:99]:
```suggestion
@pytest.mark.parametrize(
    "paid_strategy",
    [
        MarkAsPaidStrategy.TRANSACTION_FLOW,
        MarkAsPaidStrategy.PAYMENT_FLOW,
    ],
)
def test_associate_guest_checkout_with_account_if_exists_with_inactive_user(
   paid_strategy, app, address, checkout, customer_user
):
    # set the checkout email
    checkout.email = "test@example.com"
    checkout.save()
```

**Review by fowczarek — COMMENTED** [review]
I noticed one minor issue. It might be beneficial to remove `email` from the `parametrize`.

**Review by fowczarek — COMMENTED** [review]
I noticed one minor issue. It might be beneficial to remove `email` from the `parametrize`.

**Review by maarcingebala — APPROVED** [review]

**Review by fowczarek — APPROVED** [review]

**Review by szczecha — APPROVED** [review]

**Comment by maarcingebala** [comment]
Thanks for contributing to Saleor @FremahA 🎉

**Comment by maarcingebala** [comment]
BTW we opened an issue about providing the same logic in other mutation https://github.com/saleor/saleor/issues/12859.

**Comment by FremahA** [comment]
> BTW we opened an issue about providing the same logic in other mutation #12859.

Great. Kindly assign me to the new issue.

**Pull request merged**